outlines/text/sequences/sequence.py

```diff
@@ -0,0 +1,243 @@
+from typing import List, Optional, Tuple, Union
+
+import numpy as np
+from numpy.typing import NDArray
+
+
+class Sequence:
```

Comment:
Wondering if `SequenceGenerator` might be a better name. We might want to introduce a `Sequence` state object later on to track quantities such as last output, logprobs and the KV cache.

Reply:
In the same vein, should we rename the `sequences` directory to `generation`?  Calling `text.generation.completion` is more explicit than `text.sequences.completion` imo.

Reply:
Oh and should this be an abstract base class?

Reply:
> Wondering if `SequenceGenerator` might be a better name. We might want to introduce a `Sequence` state object later on to track quantities such as last output, logprobs and the KV cache.

Sounds like we might want to make these actual Python generators.

Reply:
The issue I see with making these Python generators, is that there are cases in which we need to post-process the generated string. For instance when you want to stop generation after a given string has been generated, you generally don't want said string to be included (see [here](https://github.com/normal-computing/outlines/blob/7901224b3a663ed9863a89a2d08f0744c6184781/examples/pick_odd_one_out.py#L37) for instance).

Reply:
> The issue I see with making these Python generators, is that there are cases in which we need to post-process the generated string. For instance when you want to stop generation after a given string has been generated, you generally don't want said string to be included (see [here](https://github.com/normal-computing/outlines/blob/7901224b3a663ed9863a89a2d08f0744c6184781/examples/pick_odd_one_out.py#L37) for instance).

Could we formulate this in a more functional/"stream" processing way, so that the kind of post-processing you describe is applied to the stream/generator returned to the user?

Reply:
It's worth trying